File: tools/wasmExamples/vercelWasmStream/rs.c
extern void jsSendData(int);
extern void jsAfter(void);
extern void jsBefore(void);
extern void jsWait(int);
void rs(int c)
{
    jsBefore();
    for(int i = 0; i<c; i++) {
        jsSendData(i);
        jsWait(1);
    }
    jsAfter();
}
// emcc  rs.c --no-entry -o rs.wasm  -sINITIAL_HEAP=96MB -O3 -sEXPORTED_FUNCTIONS=_rs -sERROR_ON_UNDEFINED_SYMBOLS=0 -sASYNCIFY -sASYNCIFY_IMPORTS=jsSendData -s ASYNCIFY_STACK_SIZE=290000

File: tools/wasmExamples/vercelWasmStream/rs.ts
// @ts-ignore
import wasm from "../wasm/allWASM/rs1.wasm?module";
import asyncifyThings from "./asyncifyThings";
export const config = {
  runtime: "edge",
};

export default async function handler(request: Request, event: Event) {
  let asyncifyClass;
  const encoder = new TextEncoder();
  let ctrl;
  const customReadable = new ReadableStream({
    start(controller) {
      ctrl = controller;
    },
  });
  const { exports } = (await WebAssembly.instantiate(wasm, {
    env: {
      jsBefore: () => {
        ctrl.enqueue(encoder.encode("Start"));
      },
      jsAfter: () => {
        ctrl.enqueue(encoder.encode("Done"));
        ctrl.close();
      },
      jsSendData: (i) => {
        asyncifyClass.runAsync(() => {
          ctrl.enqueue(encoder.encode(i + " "));
        });
      },
      jsWait: (i) => {
        const a = Date.now();
        while (Date.now() < a + i * 1000);
      },
    },
  })) as any;

  asyncifyClass = new asyncifyThings(exports);
  asyncifyClass.runWasm(exports.rs, [10]);
  return new Response(customReadable, {
    headers: { "Content-Type": "text/html; charset=utf-8" },
  });
}
